// joopl-analyzer/test/test.js
// Fake
$import = {
	_dependencies: null,

	mapMany: function(dependencies) {
		this._dependencies = dependencies;
	}
};

require("./test-hierarchy/moduleinfo.js");

exports.joopl_analyzer_ns4_SomeClass_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns4.SomeClass"), "joopl.analyzer.ns4.SomeClass class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns4.SomeClass"];
	test.equal(6, dependencies.length, "6 files should be the count of detected file dependencies");
	test.equal(0, dependencies.indexOf("./test/test-hierarchy/ns2/B.js"), "ns2.B class found and it has been detected in the right order");
	test.equal(1, dependencies.indexOf("./test/test-hierarchy/ns2/A.js"), "ns2.A class found and it has been detected in the right order");
	test.equal(2, dependencies.indexOf("./test/test-hierarchy/ns1/A.js"), "ns1.A class found and it has been detected in the right order");
	test.equal(3, dependencies.indexOf("./test/test-hierarchy/ns3/C.js"), "ns1.B class found and it has been detected in the right order");
	test.equal(4, dependencies.indexOf("./test/test-hierarchy/ns1/B.js"), "ns3.C class found and it has been detected in the right order");
	test.equal(5, dependencies.indexOf("./test/test-hierarchy/ClassWithPropertiesAndInstances.js"), "ClassWithPropertiesAndInstances class found and it has been detected in the right order");

	test.done();
};

exports.joopl_analyzer_SomeClass_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.SomeClass"), "joopl.analyzer.SomeClass class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.SomeClass"];

	test.equal(5, dependencies.length, "5 files should be the count of detected file dependencies");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/A.js") == 0, "ns2.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/B.js") == 1, "ns2.B class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/A.js") == 2, "ns1.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/B.js") == 3, "ns1.B class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/SomeClass.js") == 4, "SomeClass class found and it has been detected in the right order");

	test.done();
};

exports.joopl_analyzer_ns6_SomeClass_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns6.SomeClass"), "joopl.analyzer.ns6.SomeClass class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns6.SomeClass"];

	test.equal(2, dependencies.length, "2 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/SomeEnum.js") == 0, "SomeEnum class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/SomeClassUsesTestEnum.js") == 1, "SomeClassUsesTestEnum class found and it has been detected in the right order");

	test.done();
}

exports.joopl_analyzer_ns5_TestEnum_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns5.TestEnum"), "joopl.analyzer.ns5.TestEnum class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns5.TestEnum"];

	test.equal(1, dependencies.length, "1 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/SomeEnum.js") == 0, "SomeEnum class found and it has been detected in the right order");

	test.done();
}

exports.threeclasstest1_A_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("threeclasstest1.A"), "threeclasstest1.A class has been detected");

	var dependencies = $import._dependencies["threeclasstest1.A"];

	test.equal(1, dependencies.length, "1 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/ThreeClassesDifferentNamespace.js") == 0, "threeclasstest1.A class found and it has been detected in the right order");

	test.done();
}

exports.threeclasstest2_B_hasExpectedDependencies = function(test) {
	test.ok(!$import._dependencies.hasOwnProperty("threeclasstest2.B"), "This class must not be detected (single namespace member per source file");
	test.done();
}

exports.threeclasstest3_C_hasExpectedDependencies = function (test) {
    test.ok(!$import._dependencies.hasOwnProperty("threeclasstest3.C"), "This class must not be detected (single namespace member per source file");
    test.done();
}

exports.threeclasstest_A_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("threeclasstest.A"), "threeclasstest.A class has been detected");

	var dependencies = $import._dependencies["threeclasstest.A"];

	test.equal(1, dependencies.length, "1 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/ThreeClassesSameNamespace.js") == 0, "threeclasstest.A class found and it has been detected in the right order");

	test.done();
}

exports.threeclasstest_B_hasExpectedDependencies = function (test) {
    test.ok(!$import._dependencies.hasOwnProperty("threeclasstest.B"), "This class must not be detected (single namespace member per source file");
    test.done();
}

exports.threeclasstest_C_hasExpectedDependencies = function (test) {
    test.ok(!$import._dependencies.hasOwnProperty("threeclasstest.C"), "This class must not be detected (single namespace member per source file");
    test.done();
}

exports.joopl_analyzer_ns3_C_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns3.C"), "joopl.analyzer.ns3.C class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns3.C"];

	test.equal(5, dependencies.length, "5 files should be the count of detected file dependencies");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/A.js") == 0, "ns2.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/B.js") == 1, "ns2.B class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/A.js") == 2, "ns1.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/B.js") == 3, "ns1.B class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns3/C.js") == 4, "ns3.C class found and it has been detected in the right order");

	test.done();
};

exports.joopl_analyzer_ns2_A_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns2.A"), "joopl.analyzer.ns2.A class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns2.A"];

	test.equal(1, dependencies.length, "1 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/A.js") == 0, "ns2.A class found and it has been detected in the right order");

	test.done();
};

exports.joopl_analyzer_ns2_B_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns2.B"), "joopl.analyzer.ns2.B class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns2.B"];

	test.equal(4, dependencies.length, "2 files should be the count of detected file dependencies");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/A.js") == 0, "ns2.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/A.js") == 1, "ns1.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/B.js") == 2, "ns1.B class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns2/B.js") == 3, "ns2.B class found and it has been detected in the right order");

	test.done();
};

exports.joopl_analyzer_ns1_A_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns1.A"), "joopl.analyzer.ns1.A class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns1.A"];

	test.equal(1, dependencies.length, "1 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/A.js") == 0, "ns1.A class found and it has been detected in the right order");

	test.done();
};

exports.joopl_analyzer_ns1_B_hasExpectedDependencies = function(test) {
	test.ok($import._dependencies.hasOwnProperty("joopl.analyzer.ns1.B"), "joopl.analyzer.ns1.B class has been detected");

	var dependencies = $import._dependencies["joopl.analyzer.ns1.B"];

	test.equal(2, dependencies.length, "2 files should be the count of detected file dependencies");

	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/A.js") == 0, "ns1.A class found and it has been detected in the right order");
	test.ok(dependencies.indexOf("./test/test-hierarchy/ns1/B.js") == 1, "ns1.B class found and it has been detected in the right order");

	test.done();
};
